File: macos-swiftui/Tests/TranscriptCommentTests.swift
/* The two parsers the detail page is built on.
 *
 * Both turn something yt-dlp wrote for a machine into something a person can
 * read, and both fail QUIETLY when they are wrong -- a mis-threaded comment
 * section still renders, and a transcript with the rolling duplication left in
 * still looks like a transcript. Neither throws. So they are pinned here.
 */

import XCTest
@testable import YtdlMac

final class CommentThreadingTests: XCTestCase {
    private func objects(_ json: String) throws -> [[String: Any]] {
        let data = Data(json.utf8)
        let parsed = try XCTUnwrap(try JSONSerialization.jsonObject(with: data) as? [Any])
        return parsed.compactMap { $0 as? [String: Any] }
    }

    func testThreadsRepliesUnderParents() throws {
        /* The reply comes FIRST, before the parent it belongs to. yt-dlp gives
         * no ordering guarantee, which is the whole reason this is a two-pass
         * job rather than a fold. */
        let tops = Comment.thread(try objects("""
        [
          {"id":"a.1","parent":"a","text":"reply","author":"R","timestamp":200},
          {"id":"a","parent":"root","text":"top","author":"T"},
          {"id":"a.0","parent":"a","text":"earlier","author":"E","timestamp":100}
        ]
        """))

        XCTAssertEqual(tops.count, 1)
        XCTAssertEqual(tops[0].text, "top")
        XCTAssertEqual(tops[0].replies.count, 2)
        // Replies in time order, not arrival order.
        XCTAssertEqual(tops[0].replies[0].text, "earlier")
        XCTAssertEqual(tops[0].replies[1].text, "reply")
    }

    func testRecoversParentFromCompoundID() throws {
        /* The parent field carries the compound reply id rather than the bare
         * parent. yt-dlp's reply ids are "<parent>.<reply>", so the parent is
         * recoverable from the prefix. */
        let tops = Comment.thread(try objects("""
        [
          {"id":"top1","parent":"root","text":"t","author":"A"},
          {"id":"top1.9","parent":"top1.9","text":"r","author":"B"}
        ]
        """))

        XCTAssertEqual(tops.count, 1)
        XCTAssertEqual(tops[0].replies.count, 1)
    }

    func testOrphanReplyIsPromotedNotDropped() throws {
        /* The parent is genuinely absent -- a deleted comment, or a fetch that
         * was truncated. Silently losing archived text would be the worse
         * failure, so the reply is shown at top level instead. */
        let tops = Comment.thread(try objects("""
        [{"id":"zz.1","parent":"gone","text":"orphan","author":"O"}]
        """))

        XCTAssertEqual(tops.count, 1)
        XCTAssertEqual(tops[0].text, "orphan")
    }

    func testPinnedFirstThenLikes() throws {
        let tops = Comment.thread(try objects("""
        [
          {"id":"a","parent":"root","text":"few","author":"A","like_count":5},
          {"id":"b","parent":"root","text":"many","author":"B","like_count":900},
          {"id":"c","parent":"root","text":"pinned","author":"C","like_count":1,
           "is_pinned":true}
        ]
        """))

        XCTAssertEqual(tops.count, 3)
        /* The order YouTube itself shows. A comment section in arbitrary order
         * is a different document from the one people actually read. */
        XCTAssertEqual(tops[0].text, "pinned")
        XCTAssertEqual(tops[1].text, "many")
        XCTAssertEqual(tops[2].text, "few")
    }

    func testEqualLikesKeepTheirOriginalOrder() throws {
        /* Swift's sort is not stable, so equally-liked comments -- most of a
         * long thread -- need an explicit tie-break or the section reshuffles
         * itself between two reads of the same file. */
        let json = "[" + (0..<20).map {
            "{\"id\":\"c\($0)\",\"parent\":\"root\",\"text\":\"\($0)\",\"author\":\"A\"}"
        }.joined(separator: ",") + "]"

        let tops = Comment.thread(try objects(json))
        XCTAssertEqual(tops.map { $0.text }, (0..<20).map { String($0) })
    }

    func testMissingFieldsAreTolerated() throws {
        let tops = Comment.thread(try objects("[{\"id\":\"a\"},{\"parent\":\"root\"}]"))

        XCTAssertEqual(tops.count, 2)
        /* No author is "(unknown)", not empty -- every field is optional
         * because the shape varies by extractor and by yt-dlp version. */
        XCTAssertEqual(tops[0].author, "(unknown)")
        XCTAssertEqual(tops[0].likeCount, -1)
    }

    func testEmptyCommentList() {
        XCTAssertTrue(Comment.thread([]).isEmpty)
    }

    func testTotalCountsRepliesToo() throws {
        let tops = Comment.thread(try objects("""
        [
          {"id":"a","parent":"root","text":"t","author":"A"},
          {"id":"a.1","parent":"a","text":"r","author":"B"},
          {"id":"a.2","parent":"a","text":"r2","author":"C"}
        ]
        """))
        XCTAssertEqual(Comment.totalCount(tops), 3)
    }
}

final class TranscriptTests: XCTestCase {
    private var dir = ""

    override func setUpWithError() throws {
        dir = try FixtureSupport.makeTempDir(prefix: "ytdl-macos-vtt")
    }

    override func tearDownWithError() throws {
        try? FileManager.default.removeItem(atPath: dir)
    }

    private func cues(_ vtt: String) throws -> [Cue] {
        let path = Paths.join(dir, "Subtitles.en.vtt")
        try FixtureSupport.write(vtt, to: path)
        return Transcript.cues(atPath: path)
    }

    func testBasicVTT() throws {
        let c = try cues("""
        WEBVTT

        00:00:01.000 --> 00:00:03.500
        First line

        00:00:04.000 --> 00:00:06.000
        Second line
        """)

        XCTAssertEqual(c.count, 2)
        XCTAssertEqual(c[0].text, "First line")
        XCTAssertEqual(c[0].start, 1.0, accuracy: 0.01)
        XCTAssertEqual(c[0].end, 3.5, accuracy: 0.01)
    }

    func testCollapsesRollingAutoCaptions() throws {
        /* YouTube's ASR output is a rolling two-line display: each cue repeats
         * the previous cue's text and adds a few words. Read as-is it is
         * unusable as prose -- this is the single most important thing this
         * parser does.
         *
         * THREE lines, not two, and that is the whole test. Comparing each cue
         * against the last EMITTED text instead of the previous FULL cue works
         * for two lines and then silently stops: line three shares no prefix
         * with the tail emitted for line two, so it comes out whole and the
         * duplication returns for every line after. The same defect was in
         * orchid-cobalt's src-tauri/src/archive.rs. */
        let c = try cues("""
        WEBVTT

        00:00:01.000 --> 00:00:03.000
        the quick brown fox

        00:00:03.000 --> 00:00:05.000
        the quick brown fox jumps over

        00:00:05.000 --> 00:00:07.000
        the quick brown fox jumps over the lazy dog
        """)

        XCTAssertEqual(c.count, 3)
        XCTAssertEqual(c[0].text, "the quick brown fox")
        XCTAssertEqual(c[1].text, "jumps over")
        XCTAssertEqual(c[2].text, "the lazy dog")
    }

    func testIdenticalCueExtendsRatherThanRepeats() throws {
        let c = try cues("""
        WEBVTT

        00:00:01.000 --> 00:00:02.000
        same text here

        00:00:02.000 --> 00:00:09.000
        same text here
        """)

        XCTAssertEqual(c.count, 1)
        XCTAssertGreaterThan(c[0].end, 8.99)
    }

    func testShortRepeatsAreNotChopped() throws {
        /* The 12-character floor. Without it a genuinely repeated short line is
         * cut into fragments that read as a stutter. */
        let c = try cues("""
        WEBVTT

        00:00:01.000 --> 00:00:02.000
        Yeah.

        00:00:02.000 --> 00:00:03.000
        Yeah. Right.
        """)

        XCTAssertEqual(c.count, 2)
        XCTAssertEqual(c[1].text, "Yeah. Right.")
    }

    func testStripsKaraokeTagsAndEntities() throws {
        let c = try cues("""
        WEBVTT

        00:00:01.000 --> 00:00:03.000 align:start position:0%
        <c>hello</c><00:00:01.500><c> there</c> &amp; welcome
        """)

        XCTAssertEqual(c.count, 1)
        // Tags gone, entity decoded, runs of whitespace collapsed to one space.
        XCTAssertEqual(c[0].text, "hello there & welcome")
    }

    func testTimestampForms() throws {
        /* mm:ss with a comma decimal (SRT style), and a one-digit fraction:
         * ".5" is 500ms, not 5ms, which is what padding on the right is for. */
        let c = try cues("""
        WEBVTT

        01:02,5 --> 01:04,250
        short form
        """)

        XCTAssertEqual(c.count, 1)
        XCTAssertEqual(c[0].start, 62.5, accuracy: 0.01)
        XCTAssertEqual(c[0].end, 64.25, accuracy: 0.01)
    }

    func testBlocksWithoutTimingAreSkipped() throws {
        /* The WEBVTT header, NOTE blocks and cue numbers all arrive as blocks
         * with no "-->" in them. None of them is a cue. */
        let c = try cues("""
        WEBVTT
        Kind: captions
        Language: en

        NOTE this is a comment

        1
        00:00:01.000 --> 00:00:02.000
        only real cue
        """)

        XCTAssertEqual(c.count, 1)
        XCTAssertEqual(c[0].text, "only real cue")
    }

    func testWindowsLineEndings() throws {
        /* A .vtt written on Windows would never match the blank-line block
         * separator without normalisation first. */
        let path = Paths.join(dir, "crlf.vtt")
        try FixtureSupport.write(
            "WEBVTT\r\n\r\n00:00:01.000 --> 00:00:02.000\r\nwith carriage returns\r\n",
            to: path
        )

        let c = Transcript.cues(atPath: path)
        XCTAssertEqual(c.count, 1)
        XCTAssertEqual(c[0].text, "with carriage returns")
    }

    func testMissingFileIsEmptyNotACrash() {
        XCTAssertTrue(Transcript.cues(atPath: "/definitely/not/here.vtt").isEmpty)
    }

    func testDetectsAutoGeneratedTrack() throws {
        /* The filenames cannot tell these apart -- --write-subs and
         * --write-auto-subs both land in Subtitles/ under the same base name.
         * The contents can. */
        let auto = Paths.join(dir, "auto.vtt")
        let human = Paths.join(dir, "human.vtt")
        try FixtureSupport.write(
            "WEBVTT\n\n00:00:01.000 --> 00:00:02.000 align:start position:0%\n<c>word</c>\n",
            to: auto
        )
        try FixtureSupport.write(
            "WEBVTT\n\n00:00:01.000 --> 00:00:02.000\nA written line.\n",
            to: human
        )

        XCTAssertTrue(Transcript.isAutoGenerated(atPath: auto))
        XCTAssertFalse(Transcript.isAutoGenerated(atPath: human))
    }

    func testSubtitleExtensionPredicate() {
        XCTAssertTrue(MediaExtensions.isSubtitle(".vtt"))
        XCTAssertTrue(MediaExtensions.isSubtitle(".srt"))
        XCTAssertFalse(MediaExtensions.isSubtitle(".mkv"))
        XCTAssertFalse(MediaExtensions.isSubtitle(""))
        XCTAssertFalse(MediaExtensions.isSubtitle(nil))
    }
}

File: windows-winui/YtdlWin.Tests/TranscriptCommentTests.cs
/* The transcript collapse and the comment tree.
 *
 * Both parsers fail QUIETLY when they are wrong -- a mis-threaded comment
 * section still renders, and a transcript with the rolling duplication left in
 * still looks like a transcript. Neither throws. That is exactly why they are
 * pinned here rather than trusted.
 */

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using Xunit;
using YtdlWin.Core;

namespace YtdlWin.Tests;

public sealed class TranscriptTests : IDisposable
{
    private readonly string _dir;

    public TranscriptTests() => _dir = FixtureSupport.MakeTempDir("ytdl-win-vtt");
    public void Dispose() => FixtureSupport.Delete(_dir);

    private string WriteVtt(string body, string name = "captions.vtt")
    {
        var path = Path.Combine(_dir, name);
        FixtureSupport.Write(body, path);
        return path;
    }

    /* THE TEST THAT IS THE WHOLE POINT OF THIS FILE, and it is three lines deep
     * rather than two on purpose.
     *
     * The collapse must compare each cue against the previous cue's FULL text,
     * not against what was last emitted. Emitting the tail of cue 2 puts "jumps
     * over" in the output; comparing cue 3 against THAT finds no common prefix,
     * so cue 3 is emitted whole and the duplication comes straight back on the
     * third line and every line after.
     *
     * A two-line fixture passes against the broken version. That defect was
     * real -- it shipped in orchid-cobalt's src-tauri/src/archive.rs and was
     * found by the C port's own test -- which is why this fixture is three
     * lines and why this comment exists. */
    [Fact]
    public void RollingCaptionsCollapseThreeLinesDeep()
    {
        var path = WriteVtt("""
        WEBVTT

        00:00:01.000 --> 00:00:03.000
        the quick brown fox

        00:00:03.000 --> 00:00:05.000
        the quick brown fox jumps over

        00:00:05.000 --> 00:00:07.000
        the quick brown fox jumps over the lazy dog
        """);

        var cues = Transcript.Cues(path);
        Assert.Equal(3, cues.Count);
        Assert.Equal("the quick brown fox", cues[0].Text);
        Assert.Equal("jumps over", cues[1].Text);
        Assert.Equal("the lazy dog", cues[2].Text);
    }

    /* An identical repeat extends the previous cue rather than adding a line. */
    [Fact]
    public void AnIdenticalRepeatExtendsTheCue()
    {
        var path = WriteVtt("""
        WEBVTT

        00:00:01.000 --> 00:00:03.000
        the quick brown fox jumps

        00:00:03.000 --> 00:00:06.000
        the quick brown fox jumps
        """);

        var cues = Transcript.Cues(path);
        Assert.Single(cues);
        Assert.Equal(6.0, cues[0].End, 3);
    }

    /* The 12-character floor. Without it, "Yeah." followed by "Yeah. Right."
     * is chopped into fragments -- a genuinely repeated short line is not a
     * rolling caption. */
    [Fact]
    public void ShortRepeatsAreNotChopped()
    {
        var path = WriteVtt("""
        WEBVTT

        00:00:01.000 --> 00:00:02.000
        Yeah.

        00:00:02.000 --> 00:00:03.000
        Yeah. Right.
        """);

        var cues = Transcript.Cues(path);
        Assert.Equal(2, cues.Count);
        Assert.Equal("Yeah. Right.", cues[1].Text);
    }

    /* WINDOWS: a .vtt written on this platform has CRLF endings, and on this
     * platform most of them do. Without normalising first, the blank-line block
     * separator never matches, the whole file parses as one block with no
     * timeline, and the transcript tab shows nothing at all -- with no error. */
    [Fact]
    public void CrlfLineEndingsParse()
    {
        var body = "WEBVTT\r\n\r\n00:00:01.000 --> 00:00:03.000\r\nhello there\r\n";
        var path = Path.Combine(_dir, "crlf.vtt");
        FixtureSupport.WriteBytes(Encoding.UTF8.GetBytes(body), path);

        var cues = Transcript.Cues(path);
        Assert.Single(cues);
        Assert.Equal("hello there", cues[0].Text);
    }

    /* YouTube's ASR emits per-word karaoke timestamps and <c> spans. Read as-is
     * the text is unusable as prose. */
    [Fact]
    public void InlineTagsAndEntitiesAreStripped()
    {
        var path = WriteVtt("""
        WEBVTT

        00:00:01.000 --> 00:00:03.000 align:start position:0%
        <00:00:01.200><c>rock</c> <00:00:01.800><c>&amp; roll</c>
        """);

        var cues = Transcript.Cues(path);
        Assert.Single(cues);
        Assert.Equal("rock & roll", cues[0].Text);
    }

    /* &amp; must be unescaped LAST, or "&amp;lt;" becomes "<" instead of
     * "&lt;" -- the classic ordering bug in a hand-rolled unescaper. */
    [Fact]
    public void EntitiesAreUnescapedInTheRightOrder()
        => Assert.Equal("&lt;", Transcript.UnescapeEntities("&amp;lt;"));

    [Theory]
    [InlineData("00:00:01.500", 1.5)]
    [InlineData("00:01:30.000", 90.0)]
    [InlineData("01:02:03.000", 3723.0)]
    [InlineData("01:30.000", 90.0)]
    [InlineData("00:00:01,500", 1.5)]        // SRT's comma
    [InlineData("00:00:01.5", 1.5)]          // padded RIGHT: .5 is 500ms
    [InlineData("00:00:01.05", 1.05)]
    [InlineData("00:00:02.000 align:start position:0%", 2.0)]
    public void TimestampsParse(string text, double expected)
    {
        var seconds = Transcript.Seconds(text);
        Assert.NotNull(seconds);
        Assert.Equal(expected, seconds!.Value, 4);
    }

    [Theory]
    [InlineData("not a timestamp")]
    [InlineData("90")]                       // one field is not a timestamp
    public void AMalformedTimestampYieldsNull(string text)
        => Assert.Null(Transcript.Seconds(text));

    /* A file with no readable cue is empty, not a crash. */
    [Fact]
    public void AFileWithNoCuesYieldsNothing()
        => Assert.Empty(Transcript.Cues(WriteVtt("WEBVTT\n\nnothing useful here\n")));

    [Fact]
    public void AMissingFileYieldsNothing()
        => Assert.Empty(Transcript.Cues(Path.Combine(_dir, "does-not-exist.vtt")));

    /* The filenames cannot tell an auto-generated track from a written one --
     * --write-subs and --write-auto-subs land in the same folder under the same
     * base name. The contents can. */
    [Fact]
    public void AutoGeneratedTracksAreRecognisedByTheirContents()
    {
        var auto = WriteVtt("""
        WEBVTT

        00:00:01.000 --> 00:00:03.000 align:start position:0%
        <00:00:01.200><c>hello</c>
        """, "auto.vtt");

        var written = WriteVtt("""
        WEBVTT

        00:00:01.000 --> 00:00:03.000
        hello there
        """, "written.vtt");

        Assert.True(Transcript.IsAutoGenerated(auto));
        Assert.False(Transcript.IsAutoGenerated(written));
    }
}

public sealed class CommentThreadTests
{
    private static List<System.Text.Json.JsonElement> Parse(string json)
        => JsonFile.ObjectFrom(json).Objects("comments");

    /* yt-dlp writes comments as a FLAT array with a `parent` field, IN NO
     * GUARANTEED ORDER. A reply can appear before its parent, so threading is a
     * two-pass job rather than a fold -- and this fixture puts the reply first
     * precisely so a fold would fail it. */
    [Fact]
    public void ARepliesBeforeItsParentStillThreads()
    {
        var comments = Comment.Thread(Parse("""
        { "comments": [
          { "id": "a.1", "parent": "a", "text": "the reply", "author": "B" },
          { "id": "a", "parent": "root", "text": "the parent", "author": "A" }
        ] }
        """));

        var top = Assert.Single(comments);
        Assert.Equal("the parent", top.Text);
        Assert.Equal("the reply", Assert.Single(top.Replies).Text);
    }

    /* A reply whose parent is genuinely absent -- a deleted comment, or a
     * truncated fetch -- is shown at top level rather than dropped. Silently
     * losing archived text would be the worse failure. */
    [Fact]
    public void AnOrphanIsShownRatherThanDropped()
    {
        var comments = Comment.Thread(Parse("""
        { "comments": [
          { "id": "z.1", "parent": "nobody", "text": "orphan", "author": "B" }
        ] }
        """));

        Assert.Single(comments);
        Assert.Equal("orphan", comments[0].Text);
    }

    /* yt-dlp's reply ids are "<parent>.<reply>", so the parent is recoverable
     * even when the parent field itself is unhelpful. */
    [Fact]
    public void AParentIsRecoveredFromTheReplyIdShape()
    {
        var comments = Comment.Thread(Parse("""
        { "comments": [
          { "id": "a", "parent": "root", "text": "parent", "author": "A" },
          { "id": "a.1", "parent": "a.something-odd", "text": "reply", "author": "B" }
        ] }
        """));

        var top = Assert.Single(comments);
        Assert.Single(top.Replies);
    }

    /* Pinned first, then most-liked -- the order YouTube itself shows, which
     * matters because a comment section in arbitrary order is a different
     * document from the one people read. The pinned comment here has FEW likes,
     * so a sort on likes alone would fail this. */
    [Fact]
    public void PinnedComesFirstThenMostLiked()
    {
        var comments = Comment.Thread(Parse("""
        { "comments": [
          { "id": "a", "parent": "root", "text": "popular", "like_count": 5000 },
          { "id": "b", "parent": "root", "text": "pinned", "like_count": 3, "is_pinned": true },
          { "id": "c", "parent": "root", "text": "middling", "like_count": 100 }
        ] }
        """));

        Assert.Equal(new[] { "pinned", "popular", "middling" },
                     comments.Select(c => c.Text).ToArray());
    }

    /* The tie-break on original position. List.Sort is an unstable introsort,
     * so without an explicit final comparison the order of equally-liked
     * comments -- which is most of a long thread -- would differ between two
     * reads of the same file. */
    [Fact]
    public void EquallyLikedCommentsKeepTheirOriginalOrder()
    {
        const string json = """
        { "comments": [
          { "id": "a", "parent": "root", "text": "first", "like_count": 7 },
          { "id": "b", "parent": "root", "text": "second", "like_count": 7 },
          { "id": "c", "parent": "root", "text": "third", "like_count": 7 },
          { "id": "d", "parent": "root", "text": "fourth", "like_count": 7 }
        ] }
        """;

        var once = Comment.Thread(Parse(json)).Select(c => c.Text).ToArray();
        var twice = Comment.Thread(Parse(json)).Select(c => c.Text).ToArray();

        Assert.Equal(new[] { "first", "second", "third", "fourth" }, once);
        Assert.Equal(once, twice);
    }

    /* Replies are ordered oldest first, and a reply with no timestamp sorts as
     * zero rather than being dropped. */
    [Fact]
    public void RepliesAreOrderedOldestFirst()
    {
        var comments = Comment.Thread(Parse("""
        { "comments": [
          { "id": "a", "parent": "root", "text": "parent" },
          { "id": "a.2", "parent": "a", "text": "later", "timestamp": 2000 },
          { "id": "a.1", "parent": "a", "text": "earlier", "timestamp": 1000 }
        ] }
        """));

        var replies = comments[0].Replies.Select(r => r.Text).ToArray();
        Assert.Equal(new[] { "earlier", "later" }, replies);
    }

    [Fact]
    public void MissingFieldsGetSensibleDefaults()
    {
        var comments = Comment.Thread(Parse("""{ "comments": [ { "parent": "root" } ] }"""));
        var c = Assert.Single(comments);

        Assert.Equal("", c.CommentId);
        Assert.Equal("", c.Text);
        Assert.Equal("(unknown)", c.Author);
        Assert.Equal(-1, c.LikeCount);
        Assert.Equal(-1, c.Timestamp);
        Assert.False(c.IsPinned);
    }

    [Fact]
    public void TotalCountIncludesReplies()
    {
        var comments = Comment.Thread(Parse("""
        { "comments": [
          { "id": "a", "parent": "root" },
          { "id": "a.1", "parent": "a" },
          { "id": "a.2", "parent": "a" },
          { "id": "b", "parent": "root" }
        ] }
        """));

        Assert.Equal(2, comments.Count);
        Assert.Equal(4, Comment.TotalCount(comments));
    }

    [Fact]
    public void NoCommentsIsAnEmptyListNotAFailure()
        => Assert.Empty(Comment.Thread(Parse("""{ "title": "x" }""")));
}
